### PR TITLE
Try to get completions working.

### PR DESCRIPTION
**Pull request by Carreau**
I get roughly codemirror to display hardcoded completions, i found the completions messages in `api/messaging`, 

But I'm getting stuck in the Rxjs/action/agenda layers, I'll appreciate pointers.

### CODE REVIEW

**Comment by rgbkrk** [comment]
Ok, awesome. I'll be glad to help.

**Comment by Carreau** [comment]
> Ok, awesome. I'll be glad to help.

Great, thanks.  I moved forward, I achieved to call successfully to the completion, and send a complete_request though, no complete reply seem to ever reach back.

I'll cleanup what I have an push.

File: src/notebook/api/messaging/index.js [hunk 1]
```diff
@@ -55,6 +55,7 @@ export function childOf(parentMessage) {
     // since we're in an arrow function `this` is from the outer scope.
     // save our inner subscription
     const subscription = this.subscribe(msg => {
+      console.info('looking for complete_reply...', msg.parent_header.msg_type);
```
Comment (Carreau) [inline src/notebook/api/messaging/index.js]: so here I watch for all messages, I never see a complete-reply.

File: src/notebook/api/messaging/index.js [hunk 2]
```diff
@@ -55,6 +55,7 @@ export function childOf(parentMessage) {
     // since we're in an arrow function `this` is from the outer scope.
     // save our inner subscription
     const subscription = this.subscribe(msg => {
+      console.info('looking for complete_reply...', msg.header.msg_type);
```
Comment (Carreau) [inline src/notebook/api/messaging/index.js]: Also, not only do I never see a complete_reply, but I see complete_request only after doing 1 cell execution. 

Is that normal ?

File: src/notebook/components/notebook.js [hunk 3]
```diff
@@ -5,7 +5,9 @@ import HTML5Backend from 'react-dnd-html5-backend';
 
 import DraggableCell from './cell/draggable-cell';
 import CellCreator from './cell/cell-creator';
-import { executeCell, focusNextCell, moveCell } from '../actions';
+import { executeCell, focusNextCell, moveCell, getCompletions} from '../actions';
+
+import complete from '../api/messaging/completion';
```
Comment (Carreau) [inline src/notebook/components/notebook.js:10]: I'm confused about `import complete` vs `import {complete}` and what default export are, and why sometime modules in the debugger appear as `(0, _something.attribute)`.

Comment (rgbkrk) [inline src/notebook/components/notebook.js:10]: If the module only exports using `export default function blah()` then you can `import blah from './wherever'`. If it's a named import you have to use the destructuring assignment.

File: src/notebook/api/messaging/index.js (hunk 2 above)
Comment (rgbkrk) [inline src/notebook/api/messaging/index.js]: Hmmmm.

This sounds like a case of not subscribing before sending. The channels require that you subscribe to the one you expect to receive on _before_ sending. I've probably made this more difficult with that initial promise implementation I threw in api/messaging.

Comment (rgbkrk) [inline src/notebook/api/messaging/index.js]: Where does the send occur?

File: src/notebook/components/notebook.js (hunk 3 above)
Comment (jdfreder) [inline src/notebook/components/notebook.js:10]: @Carreau I don't know if this is a valid way of describing it, but I think of it like this:

``` js
import c from './a';
```

is like the following in Python

``` python
import a as c
```

where

``` js
import {b} from './a';
```

is like the following in Python

``` python
from a import b
```

File: src/notebook/api/messaging/completion.js [hunk 4]
```diff
@@ -36,6 +36,7 @@ export default function complete(channels, code, cursorPos) {
   };
 
   const p = shell
+    .logInfo()
     .childOf(message)
     .ofMessageType('complete_reply')
```
Comment (Carreau) [inline src/notebook/api/messaging/completion.js:40]: @rgbkrk refactored to logInfo only there, the send is done on line 46 below. 

This never get any messages apparently.

File: src/notebook/components/notebook.js (hunk 3 above)
Comment (Carreau) [inline src/notebook/components/notebook.js:10]: Make sens, that was roughly what I was thinking. Still weird import syntax, especially for completion, as the completer will have a hard time figuring out what you want to import before you tell it the file or module...

**Comment by Carreau** [comment]
`https://github.com/codemirror/CodeMirror/pull/3974` will allow us to use Promises with codemirror in completions.

**Comment by Carreau** [comment]
Still no clue as of why I don't get a `completion_reply`, @jdfreder ?

**Comment by jdfreder** [comment]
Did you make sure to subscribe to the channel before sending the message?  Also, have to verified that the completion reply is getting sent by the kernel?

**Comment by Carreau** [comment]
> Did you make sure to subscribe to the channel before sending the message?

Yep:

``` js
  /// create message
  const message = createMessage('complete_request');
  message.content = {
    code,
    cursor_pos: cursorPos,
  };

  /// subscribe
  const p = shell
    .logInfo()
    .childOf(message)
    .ofMessageType('complete_reply')
    .pluck('content')
    .first()
    .toPromise();

  /// send message
  shell.next(message);
  return p;
```

I'll modify the ipykernel to see log if the completion request arrives. I assumed you had that tested somewhere.

**Comment by Carreau** [comment]
Yeah ! It works ! But I have to run.

**Comment by jdfreder** [comment]
Sweet!  I'm excited to try this! Thanks @Carreau !

File: src/notebook/components/cell/editor.js [hunk 5]
```diff
@@ -70,8 +72,11 @@ export default class Editor extends React.Component {
       this.context.dispatch(updateCellSource(this.props.id, e.target.value));
     }
   }
-
   render() {
+    const extraKeys = {"Ctrl-Space": "autocomplete"}
```
Comment (jdfreder) [inline src/notebook/components/cell/editor.js:76]: At some point we will want the keymap to handle things like this.  For now we may want to change this to tab (if we can easily, without breaking indent).

**Comment by rgbkrk** [comment]
Whoa! It's alive! 

screenshot 2016-04-22 08 46 43

Extremely well done @Carreau!

**Pull request merged**